### PR TITLE
bpo-38613: Optimize set operations of dict keys.

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue38613

### CODE REVIEW

**Comment by rhettinger** [comment]
+1 This looks like a nice improvement.

**Review by rhettinger — APPROVED** [review]

File: Objects/dictobject.c [hunk 1]
```diff
@@ -4273,14 +4281,22 @@ _PyDictView_Intersect(PyObject* self, PyObject *other)
...
+        }
+    }
+    PyObject *result = PySet_New(left);
+    if (result == NULL) {
         return NULL;
+    }
 
-    tmp = _PyObject_CallMethodIdOneArg(result, &PyId_update, other);
+    _Py_IDENTIFIER(update);
+    PyObject *tmp = _PyObject_CallMethodIdOneArg(
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c]: You could use `_PySet_Update()` here.

File: Objects/dictobject.c [hunk 2]
```diff
@@ -4165,14 +4165,22 @@ static PySequenceMethods dictkeys_as_sequence = {
 static PyObject*
 dictviews_sub(PyObject* self, PyObject *other)
 {
-    PyObject *result = PySet_New(self);
-    PyObject *tmp;
-    _Py_IDENTIFIER(difference_update);
-
-    if (result == NULL)
+    PyObject *left = self;
+    if (PyDictKeys_Check(self)) {
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c:4172]: Would it be worthwhile to move the repeated code into a function? Would not it add an overhead?

File: Objects/dictobject.c [hunk 3]
```diff
@@ -4165,14 +4165,22 @@ static PySequenceMethods dictkeys_as_sequence = {
...
+        }
+    }
+    PyObject *result = PySet_New(left);
+    if (result == NULL) {
         return NULL;
+    }
 
-    tmp = _PyObject_CallMethodIdOneArg(result, &PyId_difference_update, other);
+    _Py_IDENTIFIER(difference_update);
+    PyObject *tmp = _PyObject_CallMethodIdOneArg(
```
Comment (serhiy-storchaka) [inline Objects/dictobject.c:4191]: I am wondering whether calling `set.difference_update(result, other)` using fast call can be faster than calling `result.difference_update(other)`. But this is for other issue.

**Comment by serhiy-storchaka** [comment]
LGTM. Sorry, forgot to approve the PR.

**Pull request merged**